include/permissions.hxx

```diff
@@ -0,0 +1,277 @@
+#pragma once
+#include <array>
+#include <initializer_list>
+#include <map>
+#include <string>
+#include <vector>
+
+/// Ways in which someone is allowed to access the variable, with
+/// increasing levels of rights. "ReadIfSet" indicates that
+/// variables should only be read if already set. "Final" refers to
+/// the last time anyone is allowed to write to the variable. These
+/// two concepts need to be captured to decide the order in which to
+/// execute components.
+enum class PermissionTypes { None = -1, ReadIfSet, Read, Write, Final, END };
+
+/// \defgroup RegionsGroup
+/// @{
+
+/// The regions of the domain to which a particular permission
+/// apply. These are designed to be used as bit-flags.
+enum class Regions {
+  Nowhere = 0,
+  Interior = 1 << 0,
+  Boundaries = 1 << 1,
+  All = Interior | Boundaries
+};
+
+constexpr Regions operator&(Regions a, Regions b) {
+  return static_cast<Regions>(static_cast<int>(a) & static_cast<int>(b));
+}
+
+constexpr Regions operator|(Regions a, Regions b) {
+  return static_cast<Regions>(static_cast<int>(a) | static_cast<int>(b));
+}
+
+constexpr Regions operator~(Regions a) {
+  return static_cast<Regions>(~static_cast<int>(a));
+}
+
+/// @}
+
+/// Class to store information on whether particular variables an be
+/// read from and/or written to. These permissions can apply on
+/// particular regions of the domain.
+class Permissions {
+public:
+  Permissions() = default;
+  Permissions(Permissions& x) = default;
+  Permissions(Permissions&& x) = default;
+  Permissions& operator=(Permissions& x) = default;
+  Permissions& operator=(Permissions&& x) = default;
+
+  const static std::map<Regions, std::string> fundamental_regions;
+
+  /// Data type for storing the regions of a variable which have a
+  /// particular level of permission. Some examples can be seen below:
+  ///
+  ///     AccessRights only_read_if_set = { Regions::All, Regions::Nowhere,
+  ///                      Regions::Nowhere, Regions::Nowhere },
+  ///                  read_only = { Regions::Nowhere, Regions::All, Regions::Nowhere,
+  ///                      Regions::Nowhere },
+  ///                  write_boundaries = { Regions::Nowhere, Regions::Nowhere,
+  ///                      Regions::Boundaries, Regions::Nowhere },
+  ///                  read_and_write_everywhere = { Regions::Nowhere,
+  ///                      Regions::AllReginos, Regions::All, Regions::Nowhere },
+  ///                  final_write_boundaries_read_interior = { Regions::Nowhere,
+  ///                      Regions::Interior, Regions::Nowhere, Regions::Boundaries };
```

Comment:
Please put each declaration as a separate statement, it vastly improves the readability, which is even more important for the docs!